guardian/braze-components
# Refactor Epic reminder button to use RRCP equivalent functionality

## What does this change?
Marketing colleagues have asked us to refactor Braze banner and epic components so that they use the reminder functionality currently available on web messaging components rather than the canvas-based 30-day functionality we use with Braze.

The ask requires us to know what the user's email address is, which is not immediately available to Braze components. We achieve this by threading a `fetchEmail` function from DCR into the Braze component which we can then invoke when the user clicks on the reminder CTA.

Work to thread the `fetchEmail` function into Braze components was completed in this PR: https://github.com/guardian/braze-components/pull/423. This PR builds the new Reminder CTA button and updates the default Epic component to use it.

Related DCR PR: https://github.com/guardian/dotcom-rendering/pull/8561

## Navigating this PR
**Thread a new DCR-supplied attribute `isCodeEnvironment` into Banner/Epic components**
- src/BrazeBannerComponent.ts
- src/BrazeEndOfArticleComponent.ts

**The new reminders functionality**
- src/logic/reminders.ts
- src/utils/env.ts

**Separate the Contribution and Reminder CTA buttons into their own components**
- src/components/PrimaryCtaButton.tsx
- src/components/ContributionCtaButton.tsx
- src/components/ReminderCtaButton.tsx

**Separate the loading dots animation into its own component file**
- src/components/CtaLoadingDotsAnimation.tsx
- src/newsletterCommon/index.tsx

**Refactor the Epic component to use new reminder functionality**
- src/Epic/index.tsx
- src/Epic/index.test.tsx
- src/Epic/index.story.tsx
- src/Epic/failure.stories.tsx

All the remaining file changes are minor fixes to Storybook files

## How to test
- Create a canvas in Braze-dev which will trigger the Braze epic to be added to an article
- Merge this PR into main, publish the repo to NPM, and update the DCR PR to use the latest version
- Deploy the DCR PR to CODE
- Navigate to the Guardian CODE site, view articles until the epic appears
- Click on the Reminder CTA and confirm it works
- Check that a reminder has been created for the user in the appropriate system(s)

## Images

Before user clicks:
![Screenshot 2023-10-31 at 11 03 46](https://github.com/guardian/braze-components/assets/5357530/7f057ca8-363b-4eb2-b0c0-a57bcda27b25)

When user clicks, the flashing dots animation replaces the button while the fetchEmail and setReminder functions are in flight (not shown)

... If reminder is successfully set:
![Screenshot 2023-10-31 at 11 04 00](https://github.com/guardian/braze-components/assets/5357530/44c8d16f-a928-4dd1-8bc3-d630dd905348)

... If an error is encountered:
![Screenshot 2023-10-31 at 11 04 14](https://github.com/guardian/braze-components/assets/5357530/027f7fd0-59b7-4d5f-ab84-32a7931dd6d9)
